File: src/main/java/no/nav/aap/api/søknad/rest/AbstractWebClientAdapter.java
package no.nav.aap.api.søknad.rest;

import org.springframework.http.HttpStatus;
import org.springframework.web.reactive.function.client.ClientResponse;
import org.springframework.web.reactive.function.client.WebClient;

import java.net.URI;

import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.http.MediaType.TEXT_PLAIN;


public abstract class AbstractWebClientAdapter implements RetryAware, Pingable {
    protected final WebClient webClient;
    protected final AbstractRestConfig cfg;

    public AbstractWebClientAdapter(WebClient webClient, AbstractRestConfig cfg) {
        this.webClient = webClient;
        this.cfg = cfg;
    }

    @Override
    public void ping() {
         webClient
                .get()
                .uri(pingEndpoint())
                .accept(APPLICATION_JSON, TEXT_PLAIN)
                .retrieve()
                 .onStatus(HttpStatus::isError, ClientResponse::createException)
                 .toBodilessEntity()
                .block();
    }

    @Override
    public String name() {
        return cfg.name();
    }

    protected URI getBaseUri() {
        return cfg.getBaseUri();
    }

    @Override
    public URI pingEndpoint() {
        return cfg.pingEndpoint();
    }
}

File: src/main/kotlin/no/nav/aap/api/rest/AbstractWebClientAdapter.kt
package no.nav.aap.api.rest

import no.nav.aap.api.config.Constants
import no.nav.aap.api.util.MDCUtil
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.http.MediaType.TEXT_PLAIN
import org.springframework.web.reactive.function.client.*
import java.net.URI

abstract class AbstractWebClientAdapter(protected val webClient: WebClient, protected val cfg: AbstractRestConfig) :
    RetryAware, Pingable {

    override fun name() = cfg.name()
    protected val baseUri: URI = cfg.baseUri
    override fun pingEndpoint() = cfg.pingEndpoint()

    override fun ping() {
        webClient
            .get()
            .uri(pingEndpoint())
            .accept(APPLICATION_JSON, TEXT_PLAIN)
            .retrieve()
            .onStatus({ obj: HttpStatus -> obj.isError }) { obj: ClientResponse -> obj.createException() }
            .toBodilessEntity()
            .block()
    }

    companion object {
        fun correlatingFilterFunction() =
            ExchangeFilterFunction { req: ClientRequest, next: ExchangeFunction ->
                next.exchange(
                    ClientRequest.from(req)
                        .header(MDCUtil.NAV_CONSUMER_ID, MDCUtil.consumerId())
                        .header(MDCUtil.NAV_CALL_ID, MDCUtil.callId())
                        .header(MDCUtil.NAV_CALL_ID1, MDCUtil.callId())
                        .build()
                )
            }

        fun temaFilterFunction() =
            ExchangeFilterFunction { req: ClientRequest, next: ExchangeFunction ->
                next.exchange(
                    ClientRequest.from(req)
                        .header(Constants.TEMA, Constants.AAP)
                        .build()
                )
            }
    }

    override fun toString() = "${javaClass.simpleName} [webClient=$webClient,graphQLWebClient=$cfg]"
}